### ISSUE TITLE
[ENH] Support adjacency matrix metadata in clustering workflows

### ISSUE THREAD

**Issue by ojasarora77** [skfolio/skfolio#241]
**Is your feature request related to a problem? Please describe.**

Current clustering workflows in SKFolio derive hierarchy entirely from return/correlation-based distance estimators.

In some portfolio construction settings, additional structural relationships between assets are important but cannot currently be incorporated directly into clustering workflows.

These relationships can naturally be represented as adjacency matrices or graph metadata.

---

**Describe the solution you'd like**

Allow clustering estimators and/or distance estimators to optionally consume metadata such as adjacency matrices via sklearn metadata routing or estimator parameters.

Example:

```python
model.fit(
    X,
    adjacency_matrix=adjacency
)
```

or

```python
metadata={
    "adjacency_matrix": adjacency
}
```

This would enable:

* graph-aware hierarchical clustering
* hybrid distance construction
* topology-aware linkage logic
* more flexible research workflows

This seems aligned with SKFolio’s existing modular estimator architecture and metadata routing support.

---

**Describe alternatives you've considered**

An alternative approach is manually combining adjacency information with return-derived distance matrices before passing them into clustering workflows.

However, this pushes graph integration logic outside SKFolio’s estimator pipeline and makes experimentation with metadata-aware clustering less modular and reusable.

---

**Additional context**

The enhancement would mainly benefit workflows where relationships between assets are not fully captured by return correlations alone.

**Comment by matteoettam09** [comment]
Context here as well: 

https://www.orionfinance.ai/blog/topology-risk

https://networkx.org/documentation/stable/reference/generated/networkx.linalg.graphmatrix.adjacency_matrix.html

**Comment by HugoDelatte** [comment]
I think that would be a great addition to the library.


This should probably live as a new `GraphDistance` estimator rather than directly in HRP.

The contract would stay the same as other distance estimators: after `fit`, it exposes `codependence_` and `distance_`. HRP/HERC/NCO can then consume it through the existing `distance_estimator` param:

```python
model = HierarchicalRiskParity(distance_estimator=GraphDistance())

model.fit(X, adjacency_matrix=adjacency_matrix)
````

I would also make `GraphDistance` request `adjacency_matrix` by default with metadata routing, so users do not need to manually call:

```python
.set_fit_request(adjacency_matrix=True)
```

by setting :

```python
__metadata_request__fit = {"adjacency_matrix": True}
```


I would keep a single estimator with two graph modes. Both modes consume a square adjacency matrix, but they interpret it differently.

1. `GraphMode.ASSET`

When the graph is already asset-to-asset of shape  (n_assets, n_assets).


Example:

```text
adjacency_matrix:

          AAVE   COMP   MKR    UNI
AAVE      0.0    2.0    0.2    0.0
COMP      2.0    0.0    0.5    0.1
MKR       0.2    0.5    0.0    1.5
UNI       0.0    0.1    1.5    0.0
```

Here `adjacency_matrix[i, j]` is a direct graph relationship between two assets. It can be binary, but it does not have to be. For example, `2.0` can mean a stronger shared exposure than `0.5`.

The estimator would normalize this into a graph codependence matrix and then convert it into a distance matrix, for example:

```python
distance = sqrt(1 - codependence)
```

This mode is useful when the user has already precomputed asset-to-asset relationships: same issuer, same protocol family, same venue, same collateral group, shared liquidity pool, etc.

2. `GraphMode.DEPENDENCY`

When the graph contains both asset nodes and dependency nodes of shape  (n_nodes, n_nodes).

where:

```text
n_nodes >= n_assets
```

For this mode, I would require labeled DataFrames:

where adjacency_matrix.index and adjacency_matrix.columns are all graph node names and
X.columns are the usual asset names,

Example:

```text
X.columns = ["AAVE", "COMP", "MKR", "UNI"]

adjacency_matrix.index =
[
    "AAVE", "COMP", "MKR", "UNI",
    "ETH", "USDC", "Chainlink", "Ethereum", "Uniswap"
]
```

Example weighted edges:

```text
AAVE -> ETH        1.0
AAVE -> Chainlink  2.0
AAVE -> Ethereum   1.0

COMP -> ETH        0.8
COMP -> Chainlink  1.5
COMP -> Ethereum   1.0

MKR  -> USDC       2.0
MKR  -> Ethereum   1.0

UNI  -> ETH        1.2
UNI  -> Uniswap    2.0
UNI  -> Ethereum   1.0
```

As a square adjacency matrix, that could look like:

```text
              AAVE  COMP  MKR  UNI  ETH  USDC  Chainlink  Ethereum  Uniswap
AAVE           0.0   0.0  0.0  0.0  1.0   0.0     2.0        1.0       0.0
COMP           0.0   0.0  0.0  0.0  0.8   0.0     1.5        1.0       0.0
MKR            0.0   0.0  0.0  0.0  0.0   2.0     0.0        1.0       0.0
UNI            0.0   0.0  0.0  0.0  1.2   0.0     0.0        1.0       2.0
ETH            0.0   0.0  0.0  0.0  0.0   0.0     0.0        1.0       0.0
USDC           0.0   0.0  0.0  0.0  0.0   0.0     0.0        1.0       0.0
Chainlink      0.0   0.0  0.0  0.0  0.0   0.0     0.0        1.0       0.0
Ethereum       0.0   0.0  0.0  0.0  0.0   0.0     0.0        0.0       0.0
Uniswap        0.0   0.0  0.0  0.0  0.0   0.0     0.0        1.0       0.0
```

Then `GraphDistance` derives asset dependency profiles internally. With `max_path_length=1`, it uses direct dependencies:

```text
              ETH   USDC   Chainlink   Ethereum   Uniswap
AAVE          1.0    0.0      2.0         1.0        0.0
COMP          0.8    0.0      1.5         1.0        0.0
MKR           0.0    2.0      0.0         1.0        0.0
UNI           1.2    0.0      0.0         1.0        2.0
```

With a larger `max_path_length`, it can also include indirect paths like:

```text
Asset -> Market -> Collateral
Asset -> Market -> Collateral -> Issuer
```

For example:

```python
GraphDistance(
    graph_mode=GraphMode.DEPENDENCY,
    max_path_length=3,
    path_decay=0.5,
)
```

would include paths up to length 3, with longer paths contributing less.

`path_decay=0.5` means:

```text
path length 1: weight 1.00
path length 2: weight 0.50
path length 3: weight 0.25
```

Then asset similarity can be computed from the dependency profiles, for example with weighted Jaccard:

```python
similarity(i, j) = sum(min(profile_i, profile_j)) / sum(max(profile_i, profile_j))
```

This captures cases where two assets do not have a direct edge but are close because they share hidden dependencies such as the same collateral, oracle, issuer, bridge, venue, chain or protocol.

Proposed API

```python
GraphDistance(
    graph_mode=GraphMode.ASSET | GraphMode.DEPENDENCY,
    similarity_method=GraphSimilarity.WEIGHTED_JACCARD,
    max_path_length=1,
    path_decay=0.5,
    directed=True,
    distance_method=GraphDistanceMethod.ANGULAR,
)
```

I would keep this as a pure graph distance estimator.

If we later want to combine graph distance with Pearson/covariance distance, I would rather add a separate meta-estimator like `CombinedDistance`.

WDYT?
@matteoettam09 does that work with the input you had in mind?

**Comment by matteoettam09** [comment]
Yes, this is very close to what we had in mind. I do have a few thoughts, though.

The main one is that the dependency graph should not necessarily be represented as a square asset-by-asset adjacency matrix. In general, the DAG will contain many more nodes than there are assets in the investment universe (protocols, issuers, collateral types, oracles, chains, etc.), so I would think of the graph itself as existing independently of the asset universe. The assets are simply a subset of the graph nodes, or even just entry points into the graph.

I also like the idea of weighted edges. I actually think allowing arbitrary positive weights (or at least values in [0, 1]) is preferable to a binary adjacency matrix, since the edge weights can encode the intensity or confidence of a dependency rather than just its existence.

One thing I'm less convinced about is tying `max_path_length` directly to the graph distance estimator. I see that more as one possible way of generating the graph rather than a fundamental property of the distance itself. So it's an hyperparameter of the metadata themselves, not of the class consuming them.

More generally, I think `GraphDistance` could become a much richer abstraction than just a distance estimator. For example, imagine a setting where there is no historical return data at all, but we have a DAG with weighted dependencies and each node is assigned a prior risk. In that setting, you could use Monte Carlo or another probabilistic propagation method to project those risks onto the investment universe assets, producing a covariance or codependence structure that can then be optimized. In other words, the graph itself becomes a generative model of risk rather than simply a way of measuring pairwise similarity.

So I think there's an interesting opportunity for `GraphDistance` to support both deterministic similarity-based approaches (like the one you describe) and more generative graph-based methods in the future.

**Comment by HugoDelatte** [comment]
I think the first point to clarify is the expected input of the estimator. There are two distinct designs.

**First choice: Raw dependency graph**

For this example:

```text
Assets in the investment universe:
- AAVE
- COMP

Non-asset dependency nodes:
- ETH_COLLATERAL
- CHAINLINK_ORACLE
- ETHEREUM_CHAIN
```

The input contains the complete dependency topology:

```text
AAVE → ETH_COLLATERAL → ETHEREUM_CHAIN
AAVE → CHAINLINK_ORACLE → ETHEREUM_CHAIN
COMP → ETH_COLLATERAL → ETHEREUM_CHAIN
```

It can be represented as a square node-by-node adjacency matrix (that was my proposal above). Both axes contain all graph nodes: assets and non-asset dependencies.

```python
dependency_graph = pd.DataFrame(
    [
        # AAVE  COMP  ETH_COLLATERAL  CHAINLINK_ORACLE  ETHEREUM_CHAIN
        [0.0,   0.0,       1.0,             2.0,             0.0],  # AAVE: asset
        [0.0,   0.0,       0.8,             0.0,             0.0],  # COMP: asset
        [0.0,   0.0,       0.0,             0.0,             1.0],  # ETH_COLLATERAL: dependency
        [0.0,   0.0,       0.0,             0.0,             1.0],  # CHAINLINK_ORACLE: dependency
        [0.0,   0.0,       0.0,             0.0,             0.0],  # ETHEREUM_CHAIN: dependency
    ],
    index=[
        "AAVE",
        "COMP",
        "ETH_COLLATERAL",
        "CHAINLINK_ORACLE",
        "ETHEREUM_CHAIN",
    ],
    columns=[
        "AAVE",
        "COMP",
        "ETH_COLLATERAL",
        "CHAINLINK_ORACLE",
        "ETHEREUM_CHAIN",
    ],
)
```

or as an equivalent graph object:

```python
# Asset-to-dependency edges
graph.add_edge("AAVE", "ETH_COLLATERAL", weight=1.0)
graph.add_edge("AAVE", "CHAINLINK_ORACLE", weight=2.0)
graph.add_edge("COMP", "ETH_COLLATERAL", weight=0.8)

# Dependency-to-dependency edges
graph.add_edge("ETH_COLLATERAL", "ETHEREUM_CHAIN", weight=1.0)
graph.add_edge("CHAINLINK_ORACLE", "ETHEREUM_CHAIN", weight=1.0)
```

In this design, the estimator must project the graph onto the investment universe:

```text
raw dependency graph
→ asset dependency vectors
→ asset codependence
→ asset distance
```

This requires defining traversal and aggregation semantics, including maximum path length, path decay, multiple paths, cycles, and edge-weight composition.

For example:

```python
GraphDistance(
    max_path_length=2,
    path_decay=0.5,
)
```

could derive the following dependency vector for the asset `AAVE`:

```text
ETH_COLLATERAL       1.0
CHAINLINK_ORACLE     2.0
ETHEREUM_CHAIN       1.5
```

The `ETHEREUM_CHAIN` value is derived from the two indirect paths:

```text
AAVE → ETH_COLLATERAL → ETHEREUM_CHAIN
AAVE → CHAINLINK_ORACLE → ETHEREUM_CHAIN
```

**Second choice: Projected asset-by-dependency matrix**

For this representation:

```text
Rows:
- assets in the investment universe

Columns:
- non-asset dependency variables
```

The input is already a weighted asset-by-dependency biadjacency matrix:

```python
dependency_matrix = pd.DataFrame(
    [
        [1.0, 2.0, 1.5],
        [0.8, 1.5, 0.4],
        [0.0, 0.0, 1.0],
    ],
    index=[
        "AAVE",  # asset
        "COMP",  # asset
        "MKR",   # asset
    ],
    columns=[
        "ETH_COLLATERAL",    # dependency
        "CHAINLINK_ORACLE",  # dependency
        "ETHEREUM_CHAIN",    # dependency
    ],
)
```

Here, all graph traversal and aggregation have already been performed upstream.

The estimator only computes:

```text
asset-by-dependency matrix
→ asset codependence
→ asset distance
```

For example:

```python
model = HierarchicalRiskParity(
    distance_estimator=DependencyDistance()
)

model.fit(
    X,
    dependency_matrix=dependency_matrix,
)
```

Using weighted Jaccard similarity, this could produce:

```text
codependence_:

          AAVE   COMP    MKR
AAVE     1.000  0.600  0.222
COMP     0.600  1.000  0.121
MKR      0.222  0.121  1.000
```

followed by:

```python
distance_ = np.sqrt(1 - codependence_)
```

In this design, parameters such as `max_path_length` and `path_decay` do not belong to the distance estimator because the projection has already been resolved.

So the main design decision is whether the estimator consumes the raw dependency graph or the already projected asset-by-dependency matrix.

WDYT?

**Comment by dcelisgarza** [comment]
I'm working on a similar feature in my julia library, though it's not exactly the same approach. This issue came up while I was researching if this has been done before (hence the mentions). If you like I can share what I come up with.

**Comment by HugoDelatte** [comment]
@dcelisgarza yes please feel free to share what you find. Always good to compare notes between libs.

**Comment by dcelisgarza** [comment]
Will do. It is turning out quite a bit different but maybe both could be their own estimator/flavour.

**Comment by dcelisgarza** [comment]
Ok so I'm back after quite some time. Turns out I needed to implement a bunch of stuff before this could work with my API. I took a different approach to a path length and path decay. I simply used row-wise distances starting from the biadjacency matrix. I defaulted to angular distance, but I'm not sure that's the right call. Over in Julia-land we have Distances.jl, so my solution supports any subtype of theirs that meets the interface, so it's not that big of a deal.

```math
\begin{align}
d_{i,\,j} &= \dfrac{1}{\pi}\arccos\left(\dfrac{\boldsymbol{z}_{i} \cdot \boldsymbol{z}_{j}}{\lVert\boldsymbol{z}_{i}\rVert \lVert\boldsymbol{z}_{j}\rVert}\right)\,,
\end{align}
```

Where:
  
  - $d_{i,\,j​}$: Pairwise distance between assets ``i`` and ``j``.
  - $z_i$​: Feature vector of asset ``i``, its row of the feature matrix.

Honestly it's a very different solution to what you envision, but it requires no tuning of the path length or decay. I'd probably call it something else than graph distance. It just compares how similar rows of the feature/adjacency matrix are to each other.
